Track vector disk usage by vectorReader.getOffHeapByteSize

mayya-sharipova opened the pull request:
Currently IndexDiskUsageAnalyzer reports disk usage of vectors by 
- Iterating through document values to access vector data
- Performing sample searches to force loading of the index structures
- using a sampling approach (only visiting a subset of documents based on log scale)
- tracking all bytes read during these operations

One problem of this approach is that it is very slow.
Another problem is that modifications to search algorithms and different encodings make it difficult to write definite test and assert expected results, hence a test failure such as #127689.

This modifies IndexDiskUsageAnalyzer for vectors to use a new introduced in Lucene 10.3 method vectorReader.getOffHeapByteSize. As all vector files are offHeap, we can rely on this method to report the precise disk
usage.

Closes #127689

mayya-sharipova commented:
@ChrisHegarty This is a draft PR to test the correctness of this statement: "can we can rely on vectorReader.getOffHeapByteSize to report the correct disk usage"?

ChrisHegarty commented:
> @ChrisHegarty This is a draft PR to test the correctness of this statement: "can we can rely on vectorReader.getOffHeapByteSize to report the correct disk usage"?

Absolutely. The numbers reported by getOffHeapByteSize neglect the file format header/footers, etc, but they are minimal and not currently accounted for in the current implementation anyway. So I think you got it right here @mayya-sharipova. 👍

ChrisHegarty reviewed (approved):
LGTM

mayya-sharipova committed:
Track vector disk usage by vectorReader.getOffHeapByteSize

Currently IndexDiskUsageAnalyzer reports disk usage of vectors by
- Iterating through document values to access vector data
- Performing sample searches to force loading of the index structures
- using a sampling approach (only visiting a subset of documents based on log scale)
- tracking all bytes read during these operations

One problem of this approach is that it is very slow.
Another problem is that modifications to search algorithms and different encodings make it difficult to write definite test and assert expected results, hence a test failure such as #127689.

This modifies IndexDiskUsageAnalyzer for vectors to use a new introduced in Lucene 10.3 method vectorReader.getOffHeapByteSize. As all vector files are offHeap, we can rely on this method to report the precise disk
usage.

Closes #127689

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

mayya-sharipova commented:
@elasticsearchmachine "Check labels"

mayya-sharipova committed:
Adjust tests for index usage

mayya-sharipova committed:
Iteration

mayya-sharipova commented:
@ChrisHegarty Can you please review the recent changes. Thanks a lot

mayya-sharipova committed:
Merge branch 'lucene_snapshot' into IndexDiskUsageAnalyzer_vectors
